internal/service/cluster_mapping.go

```diff
@@ -0,0 +1,119 @@
+package service
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"reflect"
+
+	merrors "github.com/RedHatInsights/insights-operator-gathering-conditions-service/internal/errors"
+	"github.com/blang/semver/v4"
+	"github.com/rs/zerolog/log"
+)
+
+type ClusterMapping [][]string
+
+// IsValid check the list is in order (based on the versions), that the versions
+// can be parsed and that the remote configurations are accessible
+func (cm ClusterMapping) IsValid(remoteConfigurationPath string) bool {
+	versions := []semver.Version{} // used to check if it's sorted
+
+	if len(cm) == 0 {
+		log.Error().Interface("raw", cm).Msg("Cluster map needs to contain at least one pair of version and filepath")
+	}
+
+	for _, slice := range cm {
+		if len(slice) != 2 {
+			log.Error().Int("len", len(slice)).Strs("slice", slice).Msg("Unexpected slice length")
+			return false
+		}
+		version := slice[0]
+		versionParsed, err := semver.Make(version)
+		if err != nil {
+			log.Error().Str("version", version).Msg("Invalid semver")
+			return false
+		}
+		versions = append(versions, versionParsed)
+
+		filepath := slice[1]
+		fullFilepath := fmt.Sprintf("%s/%s", remoteConfigurationPath, filepath)
+		if _, err := os.Stat(fullFilepath); errors.Is(err, os.ErrNotExist) {
+			log.Error().Str("filepath", fullFilepath).
+				Msg("Remote configuration filepath couldn't be accessed")
+			return false
+		}
+	}
+
+	sortedVersions := make([]semver.Version, len(versions))
+	copy(sortedVersions, versions)
+	semver.Sort(sortedVersions)
+	if !reflect.DeepEqual(sortedVersions, versions) {
+		log.Error().Interface("sortedVersions", sortedVersions).
+			Interface("versions", versions).
+			Msg("Cluster mapping is not sorted")
+		return false
+	}
+
+	return true
+}
+
+// GetFilepathForVersion iterates over the cluster map returning the first
+// filepath corresponding to the ocp version. Example:
+/*
+[
+	["1.0.0", "first.json"],
+	["2.0.0", "second.json"],
+	["3.0.0", "third.json"]
+] */
+// would return first.json for versions between 1.0.0 and 2.0.0, second.json
+// for versions between 2.0.0 and 3.0.0 and third.json for versions greater
+// than 3.0.0
+func (cm ClusterMapping) GetFilepathForVersion(ocpVersionParsed semver.Version) (string, error) {
+	// check the version is not greater than the first slice
+	firstVersion, err := semver.Make(cm[0][0])
+	if err != nil {
+		log.Error().Str("version", firstVersion.String()).Err(err).Msg("Invalid semver")
+		return "", err
+	}
+
+	comparison := ocpVersionParsed.Compare(firstVersion)
+	if comparison < 0 {
```

Comment:
why not returning some empty value or something else we agree on rather than an error?

Reply:
I'd check with @tremes and @jholecek-rh . I understood the error handling would be done on IO side

Reply:
What would be the motivation for returning a config for invalid requests?

1. We mess up the mapping - this can be prevented by CI. If a messed up mapping slips through despite the CI, we will want to be able to detect that. Error responses will be more straightforward to detect, IMHO.
2. Insights Operators will send requests will invalid versions (not a semver, version completely out of range...). If an Insights Operator fails to download a config, it should [use its built-in default](https://github.com/openshift/enhancements/blob/master/enhancements/insights/rapid-recommendations.md#remote-configuration-download). I think that using the built-in default is better than using an empty config in this case.

Either way, I think the service should respond to invalid requests with errors.

As for the error responses, would it make sense to distinguish Not Found and Bad Request?

* Not a valid semver -> Bad Request
* Valid semver smaller than the first version in the mapping -> Not Found

Reply:
Sounds good. That can help IO to know when to retry a download or not.

> version completely out of range

Is that an option? I mean, if the criteria is that the last interval is `version in file <= X` then I think even 5.0 should match there. Or we want version to stick in major? (I think it would make sense, but we need to agree on that).

> I think the service should respond to invalid requests with errors.

Sounds good.

Reply:
> As for the error responses, would it make sense to distinguish Not Found and Bad Request?
> Not a valid semver -> Bad Request
> Valid semver smaller than the first version in the mapping -> Not Found

I would leave this on IO to decide. From our PoV it's not a big deal.

> Is that an option? I mean, if the criteria is that the last interval is version in file <= X then I think even 5.0 should match there.
Be careful @ikerreyes : the first line in the cluster mapping is the minimum OCP version we accept. For a cluster map looking like this:
```
[
	["1.0.0", "first.json"],
	["2.0.0", "second.json"],
	["3.0.0", "third.json"]
]
```
clusters with versions between 1.0.0 and 2.0.0 would receive first.json, second.json for versions between 2.0.0 and 3.0.0 and third.json for versions greater than 3.0.0. However, clusters with versions 0.5.0 will receive a 400 response! :)

Reply:
Yes, I know. In that case the `version in file` will be `3.0.0` (last interval) and `X = 5.0.0`.

> However, clusters with versions 0.5.0 will receive a 400 response! :)

404 if we return a NotFound, right?

Reply:
> Yes, I know. In that case the version in file will be 3.0.0 (last interval) and X = 5.0.0.

Sorry I thought X=0.5.0

> 404 if we return a NotFound, right?

I meant with the current version, not with @jholecek-rh 's proposal

Reply:
I think we're on the same page now. Please ask again if there are still things to clarify.